### DISCUSSION TITLE
FFI shared lib imports

### DISCUSSION THREAD

**Discussion by andykais** [denoland/deno#15674] [Ideas]
given that deno now supports npm imports with `npm:<package_name>`. Would it be feasible to support deno importing ffi shared libraries somehow? Something along the lines of
```ts
import { path } from 'https://deno.land/x/my-lib/native-binaries/sqlite3.so?type=ffi'

const lib = Deno.dlopen(path, symbols)
const ptr = lib.sqlite3_libversion()
console.log(Deno.core.ops.op_ffi_cstr_read(ptr))
```

Heck if we could come up with a standard pattern for providing symbols, then we could make opening ffi files even simpler:
```ts
import { path, lib } from 'https://deno.land/x/my-lib/native-binaries/sqlite3.so?type=ffi'

const ptr = lib.sqlite3_libversion()
console.log(Deno.core.ops.op_ffi_cstr_read(ptr))
```

Just spitballing, but deno could essentially look for a directory that looks like
```
my-native-folder/
  deno_symbols.ts
  sqlite3.so
  sqlite3.dylib
  sqlite3
```

that last suggestion is probably a bit too magical, but just throwing out a few ideas on how to streamline using deno libs with ffi right now. For the most part, your current options are:
- A. run an installation script to get your binary downloaded locally
- B. pass along `--allow-net` to let the script fetch a resource manually
- C. hope that there is an existing binary on the user's system

**Comment by andykais** [comment]
I will admit this _does_ get into the ugly world of file loaders and general asset management, but I think thats exactly what we need here, management of resources required to run a deno app. The advantage to putting this behind some kind of ffi specific flag in the import specifier is that we can at least restrict downloading random files to when `--allow-ffi` is enabled. The specific protocol matters less to me. Any of these seem fine:
```ts
import { path } from 'https://deno.land/x/native/sqlite3?type=ffi'
import { path } from 'ffi:https://deno.land/x/native/sqlite3'
import { path } from 'ffi!https://deno.land/x/native/sqlite3'

**Comment by andykais** [comment]
another suggestion, would be for a `deno.json` file to declare a static file manifest, though in general it seems that we do not want remote imports to rely on a deno.json file. An example though:
```json
{
  "tasks": {},
  "manifest": [
    { "path": "native-binaries/sqlite.so" },
    { "path": "native-binaries/sqlite.dylib" },
    { "path": "native-binaries/sqlite" }
  ]
}
```

**Comment by SrGeneroso** [comment]
hey? What happened to this issue? Did it go anywhere?
